tools/mpremote: Add automatic reconnection feature

andrewleech opened the pull request:
### Summary
This PR adds automatic reconnection capability to mpremote and makes it the default behavior. When a MicroPython device disconnects, mpremote automatically reconnects to the same device, eliminating the need to manually restart the tool during development.

Reconnection is now enabled by default. Users can opt-out via:
- The `once` command for per-invocation control
- `MPREMOTE_RECONNECT` environment variable (supports 0/1/true/false/yes/no/on, case-insensitive)
- Config file setting (`reconnect = False` in `~/.config/mpremote/config.py`)
- Priority: `once` command > env var > config file > default (enabled)

Key features:
- Always reconnects to the exact same port that was previously connected when originally started in `connect auto` mode
- Always reconnects to the exact same device (on any port) when started in `connect id:abc123` mode
- Preserves resume state through reconnections
- Handles Windows COM port timing issues with retry logic

Usage:
```bash
# Reconnect is now the default
mpremote repl

# Disable reconnect for this invocation
mpremote once repl

# Disable via environment variable
MPREMOTE_RECONNECT=0 mpremote repl

# Disable via config file
echo "reconnect = False" >> ~/.config/mpremote/config.py
```

This builds on the existing disconnect handling improvements that were previously merged.

### Testing
Manual testing was performed on Linux with Pyboard D and Windows 11 with ESP32S2 devices. Tests used scripted USB power cycling (via uhubctl) to simulate disconnects and verify reconnection behavior. Testing covered default auto-reconnect, the `once` command opt-out, environment variable control (including priority override behavior where `once` takes precedence over `MPREMOTE_RECONNECT=1`), auto mode port-specific reconnection, ID mode device tracking across port changes, multiple device scenarios, resume state preservation, and environment variable parsing with various formats.

### Trade-offs and Alternatives
Making reconnect the default is a breaking change for scripts that expect mpremote to exit on disconnect. These scripts will need to use the `once` command or set `MPREMOTE_RECONNECT=0`. However, the new default is more intuitive for the primary use case (interactive development).

Design decisions:
- Default changed based on maintainer feedback
- Always reconnects to the exact same port (even in auto mode) rather than any available device, prioritizing safety in multi-device environments
- Added retry logic with delays specifically for Windows COM port timing issues
- Multiple opt-out mechanisms with clear priority order provide flexibility for different use cases
- `once` command takes highest priority as it's the most explicit user intention

The code size increase from the reconnect logic is justified by the significant improvement in developer workflow.

Josverl commented:
Thanks for the effort, Ill 
> Implemented as a standalone command rather than a --reconnect flag for consistency with other mpremote commands like resume

I have never liked that I find myself needing to type 'resume' to 95% of the commands. 
I think that that resume and reconnect should be the default state, as that is what most users will expect.
 - I think the main use for soft resetting before each command is automated testing, and there adding a reset is simple.  But changing that behavior will need some thought, as it may be a breaking change for tools that depend on it ( although the docs say they should not) 

But for this very welcome reconnect behavior I would propose this to be the default, and to be opted out via `--no-reconnect/--nrc` or environment variable.

andrewleech commented:
Thanks @Josverl for bringing up the defaults! 

Yes I too have thought it would be better to have `resume` the default, but not entirely sure of what the range of user effects this might have. I do think resume would be less confusing for many people, with an optional `reset` to be used to invert that.

I'm certainly interested in this new reconnect feature being on by default, I expect to use it as _my_ default by adding it to the local config file etc at the very least. I'd certainly be happy to make that change if we get such a concensus from the maintainers!

Josverl commented:
> adding it to the local config file etc at the very least.

AFAIK that is still broken windows - #9573 would be needed for that

Josverl commented:
As I usually have multiple boards connected I want mpremote to (re)connect to a specific one.
I find the CLI for that different to my initial expectations.

| Use Case| connect syntax| Expected | Actual | |
|--------|--------|--------|--------| -
| auto| `mpremote` | `mpremote reconnect` | `mpremote reconnect` | 👌
| COM3:| `mpremote com3` | `mpremote reconnect com3` | `mpremote com3 reconnect` | 😕
| COM8:| `mpremote connect COM8` | `mpremote reconnect COM8` | `mpremote connect com8 reconnect` | 😕😕

Similar on linux

projectgus commented:
@andrewleech Changing the defaults like this sounds very sensible to me! Feel free to request a review from me after you've had time to update.

pi-anl committed:
mpremote/console: Fix Windows disconnect detection in waitchar.

Wrap the Windows ConsoleWindows.waitchar() polling loop with
exception handling for SerialException so that a device disconnect
during waitchar raises a clean "Device disconnected" error instead
of an unhandled ClearCommError traceback.

Signed-off-by: Andrew Leech <andrew.leech@planetinnovation.com.au>

pi-anl committed:
mpremote: Add automatic reconnection feature.

When a device disconnects during a repl session, mpremote now
automatically waits for the device to reappear and reconnects.
Reconnection is enabled by default and can be disabled with
--no-reconnect. For id:-based connections the device is found
on any port; for direct port connections only the same port is
retried. A new config file option "reconnect" controls the default.

Signed-off-by: Andrew Leech <andrew.leech@planetinnovation.com.au>
